# Publication Scripts - under construction !! #

## Analyses ##

*EQTL*
- matrixQTL_Script_dj.git.R  [- R script](/R.scripts/matrixQTL_Script_dj.git.R) - matrix eQTL analysis
- PEER.Apr.UK.git.R [- R script](/R.scripts/PEER.Apr.UK.git.R) - calculation of PEER covariates

*WHOLE EXON PROGNOSIS*
- ccnd1_in_500_exomes.R [- R script](/R.scripts/ccnd1_in_500_exomes.R) - splice variant calls from exome sequencing data
- correlationCalc8.R [- R script](/R.scripts/correlationCalc8.R) - neutral evolution signature

*OS/PFS GWAS*
- covariant.selection.OS.GWAS.git.r [- R script](/R.scripts/covariant.selection.OS.GWAS.git.r) - covariates selection for OS/PFS GWAS

*DNA METHYLATION*
- CL.Limma.Response.R [- R script](/R.scripts/CL.Limma.Response.R) - 
- Minfi.Dec.CL.17.2016.R [- R script](/R.scripts/Minfi.Dec.CL.17.2016.R) - QC pipelines to generate Beta-values
- sur.loop.dv.in.R [- R script](/R.scripts/sur.loop.dv.in.R) - survival loop for methylation probes

## Data Visualization ##
 
*DNA METHYLATION*
- Figure.Ex1.methOS.SurvivalROC3.git.R [- R script](/R.scripts/Figure.Ex1.methOS.SurvivalROC3.git.R)
    <img src=/figure3.svg width="1100">  
- Figure3.methOS.ForestPlot_ExpMeis2.git.R [- R script](/R.scripts/Figure3.methOS.ForestPlot_ExpMeis2.git.R)
    <img src=/R.scripts/images/Fig3.paper.jpgs.jpg width="800">
- Figure5.methOS.3.Survminer.18.git.R [- R script](/R.scripts/Figure5.methOS.3.Survminer.18.git.R)
    <img src=/figure4.svg width="550">

*NEUTRAL EVOLUTION*
- Response.neu.fig.R [- R script](/R.scripts/Response.neu.fig.R)
    <img src=/figure3.svg width="1100">  
- Vioplot.tnx.hrd.1q.CoMMpass.R [- R script](/R.scripts/Vioplot.tnx.hrd.1q.CoMMpass.R)
    <img src=/figure3.svg width="1100">  
- Surv.neu.mk.commpass.R [- R script](/R.scripts/Surv.neu.mk.commpass.R)
<img src=/figure3.svg width="1100"> 
- GVIS.commpass.R [- R script](/R.scripts/GVIS.R)
<img src=/R.scripts/images/Gvisplot.Jan31.png width="800">  